astropy/modeling/core.py

```diff
@@ -1102,10 +1102,27 @@ def _initialize_parameter_values(self, params):
 
         self._param_metrics = param_metrics
         self._parameters = np.empty(total_size, dtype=np.float64)
+
         # Now set the parameter values (this will also fill
         # self._parameters)
+        # TODO: This is a bit ugly, but easier to deal with than how this was
+        # done previously.  There's still lots of opportunity for refactoring
+        # though, in particular once we move the _get/set_model_value methods
+        # out of Parameter and into Model (renaming them
+        # _get/set_parameter_value)
         for name, value in params.items():
-            setattr(self, name, value)
+            param_descr = getattr(self, name)
+            value = np.array(value)
+            if param_descr._setter is not None:
+                value = param_descr._setter(value)
+            self._parameters[param_metrics[name]['slice']] = value.ravel()
+
+        # Finally validate all the parameters; we do this last so that
+        # validators that depend on one of the other parameters' values will
+        # work
+        for name in params:
+            param_descr = getattr(self, name)
+            param_descr.validator(param_descr.value)
```

Comment:
I see why you "set all values, then validate" here.  Assuming one wants to change the parameters after initialization, is there still a way to do that -- if the validations were cyclical in such a way that they couldn't be set and validated one at a time?  I suspect the easiest way out would be to add a `set_params` method that would take kwargs:

```
model.set_params(mu=0.1, gamma=0.25)
```

I suppose this is all academic at this point, since none of the existing validators appear to have that problem -- but the example you gave in the head of this PR might.

I had similar problems in ASDF where sometimes it was impossible to mutate the tree in such a way that each step along the way would also be valid.  I ended up just validating on write.